src/components/tl/ucp/allgather/allgather_knomial.c

```diff
@@ -234,6 +262,183 @@ ucc_status_t ucc_tl_ucp_allgather_knomial_start(ucc_coll_task_t *coll_task)
+                peer_dist = ucc_knomial_calc_recv_dist(size - p->n_extra,
+                        ucc_knomial_pattern_loop_rank(p, peer), p->radix, 0);
+                if (peer_dist < task->allgather_kn.recv_dist) {
+                    continue;
+                }
+            }
+            mmap_params.address     = sbuf;
+            mmap_params.length      = local_seg_count * dt_size;
+            status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, &mh);
+            if (UCC_OK != status) {
+                task->super.status = status;
+                return;
+            }
+            if (count_mh == size_of_list){
+                size_of_list *= 2;
+                mh_list = (ucp_mem_h *)realloc(mh_list, size_of_list * sizeof(ucp_mem_h));
+            }
+            mh_list[count_mh++] = mh;
+        }
+
+        for (loop_step = 1; loop_step < radix; loop_step++) {
+            peer = ucc_knomial_pattern_get_loop_peer(p, rank, loop_step);
+            if (peer == UCC_KN_PEER_NULL)
+                continue;
+            ucc_kn_ag_pattern_peer_seg(peer, p, &peer_seg_count,
+                                       &peer_seg_offset);
+
+            if (coll_task->bargs.args.coll_type == UCC_COLL_TYPE_BCAST) {
+                peer_dist = ucc_knomial_calc_recv_dist(size - p->n_extra,
+                        ucc_knomial_pattern_loop_rank(p, peer), p->radix, 0);
+                if (peer_dist > task->allgather_kn.recv_dist) {
+                    continue;
+                }
+            }
+            mmap_params.address     = PTR_OFFSET(rbuf, peer_seg_offset * dt_size);
+            mmap_params.length      = peer_seg_count * dt_size;
+            status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, &mh);
+            if (UCC_OK != status) {
+                task->super.status = status;
+                return;
+            }
+            if (count_mh == size_of_list){
+                size_of_list *= 2;
+                mh_list = (ucp_mem_h *)realloc(mh_list, size_of_list * sizeof(ucp_mem_h));
+            }
+            mh_list[count_mh++] = mh;
+        }
+        ucc_kn_ag_pattern_next_iter(p);
+    }
+
+    if (KN_NODE_PROXY == node_type) {
+        mmap_params.address     = args->dst.info.buffer;
+        mmap_params.length      = data_size;
+        status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, &mh);
+        if (UCC_OK != status) {
+            task->super.status = status;
+            return;
+        }
+        if (count_mh == size_of_list){
+            size_of_list *= 2;
+            mh_list = (ucp_mem_h *)realloc(mh_list, size_of_list * sizeof(ucp_mem_h));
+        }
+        mh_list[count_mh++] = mh;
+    }
+
+out:
+    ucc_assert(UCC_TL_UCP_TASK_P2P_COMPLETE(task));
+    task->super.status = UCC_OK;
+    task->mh_list = mh_list;
+    task->count_mh = count_mh-1;
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_allgather_kn_done", 0);
```

Comment:
is this needed here?

Reply:
I don't know what this is doing, do you know ?